---
title: "mono_heart_spikeIn.Rmd"
output: html_document

// 2.13.18
// Lets do this for all NE cells (fetal & adult) + 50 random heart cells spiked in from the MACA cell-
//    counts table for heart. 
// Working!!
---

Load library
```{r}
library(monocle)
```

Load data frames
```{r}
raw_data_SI <- read.csv('./heart_spikeIn/spike_in_counts.csv', header=TRUE)
row.names(raw_data_SI) <- raw_data_SI[,1]
raw_data_SI <- raw_data_SI[ ,-1]

meta_data_SI <- read.csv('./heart_spikeIn/spike_in_meta.csv', header=TRUE)
row.names(meta_data_SI) <- meta_data_SI[,1]
meta_data_SI <- meta_data_SI[ ,-1]
```

Sanity check
```{r}
dim(raw_data_SI)
dim(meta_data_SI)
```

Another check
```{r}
#for(i in 1:length(colnames(raw_data_SI))){
#  bool <- any(is.nan(raw_data_SI[,i]))
#  print(bool)
#}
```

Create annotated data frame from meta_data
```{r}
meta_data_anno_SI <- new("AnnotatedDataFrame", data = meta_data_SI)
```

Create CellDataSet object
```{r}
cds_SI <- newCellDataSet(cellData = as.matrix(raw_data_SI), phenoData = meta_data_anno_SI, featureData = NULL, expressionFamily = negbinomial.size())
```

Estimate size and dispersion
```{r}
cds_SI <- estimateSizeFactors(cds_SI)
cds_SI <- estimateDispersions(cds_SI)
```


CONSTRUCTING SINGLE CELL TRAJECTORIES


1) Choosing genes that define progress
      Using dpFeature technique
```{r}
cds_SI <- detectGenes(cds_SI, min_expr = 0.1)
fData(cds_SI)$use_for_ordering <- fData(cds_SI)$num_cells_expressed > 0.05 * ncol(cds_SI)
```


Find genes expressed in >10 cells
    Does this step work??
```{r}
all_SI_expressed_genes <- row.names(subset(fData(cds_SI), num_cells_expressed >= 10))
```

Plot PC variance
```{r}
plot_pc_variance_explained(cds_SI, return_all = F, verbose = T, max_components = 100) 
```

Run dimensionality reduction
```{r}
cds_SI <- reduceDimension(cds_SI, max_components = 2, norm_method = 'log', num_dim = 20, reduction_method = 'tSNE', verbose = T)
```

Cluster cells & plot
```{r}
cds_SI <- clusterCells(cds_SI, verbose = F)
plot_cell_clusters(cds_SI)
```

Refine clustering params
```{r}
#plot_rho_delta(cds_SI, rho_threshold = 2, delta_threshold = 4 )
```

...and recluster
```{r}
#cds_SI <- clusterCells(cds_SI, rho_threshold = 2, delta_threshold = 4, skip_rho_sigma = T, verbose = F)
#plot_cell_clusters(cds_SI)
```

Differential gene expression test
```{r}
clustering_DEG_genes_SI <- differentialGeneTest(cds_SI[all_SI_expressed_genes,], fullModelFormulaStr = '~Cluster', cores = 10)
```

Select top 1000 genes
```{r}
all_SI_ordering_genes <- 
  row.names(clustering_DEG_genes_SI)[order(clustering_DEG_genes_SI$qval)][1:1000]

cds_SI <- setOrderingFilter(cds_SI, ordering_genes = all_SI_ordering_genes)
cds_SI <- reduceDimension(cds_SI, method = 'DDRTree')
cds_SI <- orderCells(cds_SI)
```

...And plot
```{r}
plot_cell_trajectory(cds_SI)
plot_cell_trajectory(cds_SI, color_by = "origin")
plot_cell_trajectory(cds_SI, color_by = "age")
```

...and again
```{r}
plot_complex_cell_trajectory(cds_SI)
plot_complex_cell_trajectory(cds_SI, color_by = 'age')
plot_complex_cell_trajectory(cds_SI, color_by = 'origin')
```

